src/Raven.Client/Http/RequestExecutor.cs

```diff
@@ -1051,16 +1052,23 @@ private Task RefreshIfNeeded(ServerNode chosenNode, HttpResponseMessage response
 
             var response = await preferredTask.ConfigureAwait(false);
 
-            if (TryGetServerVersion(response, out var serverVersion))
-                LastServerVersion = serverVersion;
+            // PERF: The reason to avoid rechecking every time is that servers wont change so rapidly
+            //       and therefore we dimish its cost by several orders of magnitude just doing it
+            //       once in a while. 
+            if (LastServerVersion == null || _lastServerVersionCheck > (1 << 18))
```

Comment:
Huh? Why is the `1<<18` for?

Reply:
That's 32K, yeah probably better to use a constant there... I forgot that we have the Constants class to build such things.

Reply:
Or you can write 32_000 instead here, because you do not need it elsewhere.

Reply:
Also where are you resetting _lastServerVersionCheck? And 32k is too much. Consider using 128.

Reply:
We want to reduce this as much as possible. I would prefer to do a time based value.
Something like every 5 - 15 minutes or so.

In fact, I'm also fine with doing this _once_ for all a node. The chance that they'll revert back a node while the app is running is small.

Reply:
Time-based sounds like a good approach IMO.

Reply:
Time-based requires accessing the clock which is not that good. We can time it under stress but the idea is to diminish this enough so that it does not register (3 orders of magnitude is huge) --- so I changed it to 1K

Reply:
@ppekrol Why bother to refresh anyway?

Reply:
Best option would be to set `LastServerVersion` to null when we are switching nodes and then extract the server version when it is null. This would consume minimal amount of compute power. Thoughts?

Reply:
What do you mean, switching nodes?

Reply:
1. When you are talking with node A and LastServerVersion is null, set it. 
2. If you will switch for any reason to node B, set LastServerVersion to null and go to point 1.

Reply:
Won't work when we have failover, or when we use round robin, though.

Reply:
In general, going back is such a rare event, we can just do it once and be done with it.
No extra checks needed.

Reply:
The big question is: Should we want a reliability problem (even if rare) when the cost for it will become effectively 3 orders of magnitude less aka virtually 0?

Reply:
It's not `3 orders of magnitude` issue - that is micro view on the problem. From macro overview this is 1.82% (probably even less) with the reliability issue. I do not see us stepping back from path once taken. Doing that every `mod 10` call or on failover decreases the problem to neglectable level.

Reply:
I agree but right now the level of optimization of the client is from what I observed something that can be improved a lot and that will eventually will arise again, and we will need again to decide what to do. That 1.86% is a lie caused by the blocking call inside the `SendAsync` synchronization, it is roughly around 5% (still small in the grand scheme of things today) where 1 order of magnitude push it to 0.5% and 2 to 0.05% (1 in 10 vs 1 in 100).

Reply:
For some apps 100 is a lot IMO.

Reply:
@redknightlois  @ppekrol I'm missing something. This isn't a reliability issue.
This is used for one purpose, to check if the server is 4.2, which is pretty static.
We don't expect to replace the node versions (and backward) very often, so why do the additional check?

Reply:
I'm worried that server version will change and Client API will not pick it up because if we will be doing that one time then it will require the restart of the client application. Plus there is always an option to have mixed clusters during update.